### DISCUSSION TITLE
Sign up for notifications / polls / requests for comments from axxel

### DISCUSSION THREAD

**Discussion by axxel** [zxing-cpp/zxing-cpp#706] [Announcements]
Time and again, I find myself in the position to want to get in touch with "you" out there who have some stake in `zxing-cpp`. What I did in the past multiple times is compiling a list of names add hoc based on earlier contributions or random issue reports that I manage to remember. That is an inefficient strategy. To my understanding, using the github notifications feature is supposed to be used for that but that did not work very well (e.g. with the two most recent poll discussions I started). Maybe most of you did not sign up for that to not be bothered by random bug reports or ci build failures?

Anyway, new idea: use this 'meta discussion' where every participant gets notified when I post here to point your attention to a specific topic, like I'm doing right now: #705 (`Result` -> `Barcode`?).

To **subscribe**, anyone can simply post a 'hello world' comment here (I don't expect this to happen very often). Ideally you'd write a sentence about what your interests in `zxing-cpp` are, like e.g. "Author of an Android App".

To **unsubscribe**, simply press the *Unsubscribe* button on the right.

To get this started I'll do this kind of name compilation one more time:  @gitlost @markusfisch @alexmanzer @vkrause @Sec-ant @benjohnde @siiky @khoren93 @huycn @okarmazin @duanqn @SteveCookTU @stumpylog @parallaxe @ryantrem @duanqn.
EDIT: To make one thing more clear: everyone mentioned here explicitly is already *subscribed*, but if you'd still like to drop us a line about your interests, please do so. :)

In case I'm ignorant of fancy feature xyz of github that is way more appropriate for this task: "let me know in the comments down below ;)".

**Comment by gitlost** [comment]
zint testing, general interest in barcode decoding

**Comment by markusfisch** [comment]
Hi, I'm using ZXingCpp in mobile apps, Android and iOS, professionally and also in side projects. I'm interested in reading, but also very much in writing barcodes.

**Comment by khoren93** [comment]
Hello, 

I'm the author of flutter_zxing, and I appreciate your initiative to streamline communication regarding zxing-cpp. I've subscribed to this meta discussion to stay updated on relevant topics.

**Comment by Sec-ant** [comment]
Hello, I'm the maintainer of zxing-wasm, one of the WASM wrappers. I'm interested in the reading part and it's usage in web applications. I'd like to subscribe to keep information updated.

**Comment by duanqn** [comment]
Hi, I am a co-author of a commercial C++ program and we use ZXing-cpp for reading barcodes. I'm interested in improving the recognition rate for barcodes.

**Comment by vkrause** [comment]
Representing KDE here, we use ZXing via its C++ API for "conventional" barcode scanning in mobile apps, for barcode decoding in PDF documents (pure mode), and for PDF417 rendering.

I am subscribed to notifications of everything here already anyway, so I did notice the two recent discussions you mentioned, but since those don't really impact us (or as in case of the naming the impact is the same regardless of the new name) I didn't comment there.

**Comment by siiky** [comment]
I initially put the C wrapper ball rolling, which I use at work to read QR and PDF417 codes.

**Comment by axxel** [comment]
I guess everyone here has read about the new .NET wrapper somewhere by now but since I specifically set up this discussion here for this kind of information: #707  :)

**Comment by axxel** [comment]
It just so happens that already another subject requiring some community feedback came up: Please state your opinion regarding disabling of exception handling in this poll: #710.

**Comment by pranavtriguna** [comment]
Thanks for all the work ! Trying to evaluate if I may be able to use it at work. Mostly interested in C/CPP. Since the native is CPP, that bloats up my application size, trying to keep it to a minimum.
Is there a report/ experimentation with the application sizes for C/CPP builds? 
Thanks !

**Reply by axxel** [reply]
For a further discussion of this question please reply to #715.

**Comment by axxel** [comment]
The work I've just pushed in pursuit of helping detect inverted barcodes (see #711) led to a significant change in how the `LocalAverage` alias `HybridBinarizer` works. This may have a positive or negative effect on the success rate for individual images. If anyone has their own set of samples images, I'd be interested to hear about regressions and/or improvements.

**Reply by vkrause** [reply]
Our test suite seems unaffected, but that is dominated by "pure" images which I'd assume aren't really impacted by this work anyway.

**Reply by axxel** [reply]
@vkrause that is correct. If you only have black and white pixels, the old and new implementation of the `HybridBinarizer` should definitively agree. Maybe you do that already but just in case: if you know that your input is binary (generated, clean pixels without shading or noise), then please set the binarizer to `FixedThreshold`. That would be way faster and also not affected by the change at all. See [here](https://github.com/zxing-cpp/zxing-cpp/blob/e107516f8accea34dc3e4574b88186f19b06417c/example/ZXingReader.cpp#L77).

**Reply by vkrause** [reply]
Checked to be sure, we do that already indeed.

**Comment by axxel** [comment]
New poll: #722 (Should `ReaderOptions` default to the same values across all wrappers?)

**Comment by axxel** [comment]
And another poll: #725 (Mobile applications and the use of `tryRotate`). This should be of interest to everyone who has a mobile application and cares about performance / frame rate.

**Comment by axxel** [comment]
New poll: #744 (How best to disable/remove the writer or reader functionality from the build?)

**Comment by axxel** [comment]
I [changed](https://github.com/zxing-cpp/zxing-cpp/commit/21ed1e0acd50a24d4d2a6bfc6635417664060d7d) the cmake build options to use the prefix `ZXING_` instead of `BUILD_` to follow best practices. I took care to keep the scripts backward compatible with the old config names for the time being. There will be a 2.3 release and then I will likely remove the backward compatibility with the 3.0 release.

**Comment by axxel** [comment]
New poll: Add a JSON 'channel' to communicate symbology specific / niche bits of data (#862)

**Comment by axxel** [comment]
I "published" a 2.3.0-rc1 release and would be delighted if anyone interested could give it a quick check with their setup. I don't expect much reaction today and will likely do the 2.3.0 release tomorrow, because 1.1.25 sounds nice ;).

**Comment by axxel** [comment]
**Anyone interested in the WinRT wrapper?**

Since @huycn has not responded to #882, yet. I fear he might not be available to the cause of this project anymore. Is there anyone else interested in the WinRT wrapper? Since maintaining this niche windows corner while not being a windows user myself is a burden, the future of this wrapper code and accompanying NuGet package is looking bleak.

**Reply by duanqn** [reply]
Hi @axxel, I am a Windows user and I am willing to contribute, but I am relatively new to WinRT. I have read #882. Is there anything we need to do at the moment, apart from generating a new key?

**Reply by axxel** [reply]
I can't simply renew the key as I don't own the account. I could setup a new one or maybe use the one I created for the .NET wrapper. But that would basically mean, to publish a 'new' wrapper. And in that case, it would make sens to me to use the opportunity to modernize the API and bring it in line with the e.g. the .NET wrapper API. If you are interested in looking into this, please go ahead.

**Comment by axxel** [comment]
Everyone that uses pkg-config (specifically `zxing.pc`), please have a look at #921 and chime in if you have objections to the proposed change.

**Comment by axxel** [comment]
Triggered by @gitlost's progress on zint work, specifically for the `WriteBarcode` API, I'm moving that forward right now and am considering requiring c++20 for that to compile. This would mean a 3.0 release would not build without a c++20 compiler anymore. If this would break someone's use-case, please speak up and explain.

**Reply by mbfrig** [reply]
We just moved our build environments to C++17. We use VS 2022 and it defaults to C++14, so we had to manually change the projects to C++17. Not sure about doing another jump so quickly, in a production environment. What are the advantages, for ZXing, for going to C++20?

**Reply by axxel** [reply]
1. The project has had a c++20-only feature for more than 3 years now (see d6dc409ef15ff7171c7f9130dff7266a7c65f053) and people notice when this is missing: #1004). It is highly recommended using c++20 and I'd like to prevent people from running into this by forcing it.
2. There a number of `#ifdef c++20` hacks scattered around the code that I'd like to see gone.
3. Given that we are heading into 2026, I'd actually consider c++23 to be a viable standard version. I have not looked into this, though.

I appreciate your feedback but given the relative silence on the matter and my strong preference to move to (at least) c++20, I will do that for the 3.0 release. That said: there is going to be a 2.4 release first, so this will not happen this year.

**Reply by mbfrig** [reply]
Thanks for the details. I see why you want to do it. Doing the 2.4 first is a good comprise.

**Comment by axxel** [comment]
Is anyone relying on the possibility to distinguish between any of the 4 variants of the same barcode "family" EAN/UPC, namely `EAN13`, `EAN8`, `UPCA`, `UPCE`? See also [this comment](https://github.com/zxing-cpp/zxing-cpp/discussions/926#discussioncomment-12747210) from 3 years ago. They are defined in the same specification and the current implementation is technically in violation of that spec (for backward compatibility) and I'd like to fix that and thereby get rid of the distinction.

To be more specific, my proposal would
1. eliminate the option to scan e.g. only UPC-E codes
2. eliminate the possibility to discern between e.g. EAN13: 0042100005264, UPC-A: 042100005264 and UPC-E: 04252614. The spec requests that all of them are transmitted as the fully qualified string `]E00042100005264`.

**EDIT**: To clarify even more: `EAN8` will always be distinguishable from the rest via the `Barcode::symbologyIdentifier` property (`]E4` vs `]E0`). `EAN13` and `UPCA` have never been really distinguishable and just been inferred by the leading digit (0 vs non-0). That part is gone now. I'm considering deprecating the `BarcodeFormat` constants for those 4 and merge them into one new `EANUPC` (based on the term used in the spec: `EAN/UPC`). It would also be possible to save the `UPCE` info and content in the discussed JSON channel, see #862. So the first point above is still considered, the second is already (mostly) implemented.

**Reply by mbfrig** [reply]
We do allow the user to pick from a list of barcodes to detect, and we have these listed separately. It is the requirement that we have had for a long time. So we would want these to remain separate.

**Comment by axxel** [comment]
Anyone who cares about what exactly `Barcode::bytesECI()` returns, please have a look at #936.

**Comment by axxel** [comment]
New discussion/poll about the general direction of this project for the future: #1000

**Comment by mbfrig** [comment]
Another department is using an older version of this, so we are looking into using the latest in ours. We have v2.3.0 and are evaluating it.

**Comment by axxel** [comment]
Before releasing 2.4, I noticed the inconsistency between setters/getters of `ReaderOptions` vs `WriterOptions`. The former use the `set` prefix, the latter don't. I will change either of them to look like the other. We currently have `ReaderOptions::setTryRotate(bool)` and `bool ReaderOptions::tryRotate()`. I don't like the look of `WriterOptions::setAddQuietZone(bool)`.

I prefer the setters without `set` prefix and since those options structs are practically a one-way data transfer into the library where the setters are clearly more important than the getters, I'm wondering why we'd need the getters at all. Does anyone have use for them? If not, I might deprecate them and then remove them in 3.0.

EDIT: this removal would of course also apply to the C-API and consequently all wrapper APIs.

**Reply by mbfrig** [reply]
I do use a couple of the getters for ReadOptions, so I can test the defaulted values. This way I can see if any internal things changed that I need to look at to see if it would impact any of our prior testing.

**Reply by axxel** [reply]
I decided to not deprecate them (and would not have removed them in the next release anyway, independent of what version number I am attaching). I decided to even keep both the `setNAME(val)` and the `NAME(val)` setter versions as both can be considered 'natural' depending the context and one of them is simply an inline function calling the other with no extra cost in maintainability.

**Comment by axxel** [comment]
Call for testing: After realizing that incorporating the new `libzint` based backend as non-experimental would require an ABI change and hence a SONAME bump, I decided to skip the mentioned 2.4 release and go straight to 3.0. Doing that then made me want to add as many ABI breaking changes as I can foresee at the moment. This lead to the long planed switch to a PIMPL based implementation for `Barcode` and `ReaderOptions`. That in turn allowed me to remove all setters from the `Barcode` class which is now minimal. Other than the new `CreateBarcode` + `WriteBarcode` API, the `Barcode` class has a new `symbol()` and a new `extra()` member.

Bottom line: there is a lot of new code compared to 2.4 and a lot of that was added/modified in the last 3 weeks. Hence, I would highly appreciate if as many people as possible could give the current HEAD a try. And also please checkout the new writer API and speak up if something is off. There is sort of a deadline for the 3.0 release near the end of January because of the next Ubuntu LTS version.

**Reply by mbfrig** [reply]
We were expecting 2.4, so we can have all the latest fixes in our product. Changing to 3.0 is too big of a change for us to take, until the next major product update. This is unfortunate to hear.

**Reply by axxel** [reply]
May I ask what exactly it is that is "too big"? Note: the API of the former non-experimental part of the library is (supposed to be) unchanged (except for the removal of formerly deprecated features). My decision to call this 3.0 instead of 2.4 is somewhat cosmetic in nature.

**Reply by mbfrig** [reply]
In a previous comment above, back in Oct ‘25, the jump to 3.0 was also changing to require c++20. We had just move our stuff to use c++17, to support zxing; which we had to manually because VS2022 defaulted to c++14. I barely got that in at that time because of our production cycle. And because of that cycle, I can’t make the change to c++20 at this time. I understand the ‘ease’ of doing that as a developer, but there are testing team aspects that prevent it at this time.

**Reply by axxel** [reply]
I see. Surely a valid point but there is one thing that might be relevant for your setup: as soon as you have a binary compiled somehow (with a c++20 compiler), you should be able to use that from a c++17 client application. The public API of the lib stayed with c++17 as before).

**Reply by mbfrig** [reply]
Agreed. The issue is a common library that we use with binary that has zxing, is built once and shared with other modules. So it is a configuration management thing as well. Either need to builds of it or bump it up; just nothing we have time for right now. But I do have it on our roadmap, to get us to c++20.

**Reply by axxel** [reply]
@Sec-ant can you make sure the current HEAD does not have some fundamental issues in your wasm wrapper setup?

**Reply by axxel** [reply]
@vkrause I'd like to invite you to check out if the following would work for your different writer/generator use cases:
```c++
Barcode bc = CreateBarcodeFromText("Something something", BarcodeFormat::SomeFormat);
Image img = bc.symbol(); // produces a 1 module == 1 pixel black/white image
```

**Reply by axxel** [reply]
@khoren93 Please have a look at the current HEAD and check if you find any issues incorporating the new writer API into your wrapper. There is about a week left before this gets finalized.

**Reply by Sec-ant** [reply]
> @Sec-ant can you make sure the current HEAD does not have some fundamental issues in your wasm wrapper setup?

@axxel I [did a bump and run the snapshot tests](https://github.com/Sec-ant/zxing-wasm/pull/314), most of the diffs can be accounted for `extra`, `version` and `symbol` bit flip tweaks, which are expected I believe. I do noticed a strange one, the [OldDetectorFallback datamatrix-3 test case](https://github.com/zxing-cpp/zxing-cpp/blob/master/test/samples/datamatrix-3/OldDetectorFallback.jpg) yielded a different `symbol` when rotated, I think you can reproduce the issue with the following two samples:

<img src="https://github.com/user-attachments/assets/d07e0354-184c-4a18-af44-03f7f7c9a900" width="96" alt="OldDetectorFallback">

<img src="https://github.com/user-attachments/assets/40cd605b-3a88-4950-934c-d8e526dd5863" width="96" alt="OldDetectorFallback rotated">

**Reply by axxel** [reply]
@Sec-ant thanks for testing it. I had a quick look at the PR you linked and would suggest you do away with any compatibility wrapper that (apparently) copilot introduced and map the new API directly (like `forceSqare` instead of `forceSqaureDataMatrix`). The old was experimental for a reason. ;)

Regarding the different symbols for the linked DataMatrix: that came from 907e6746320d89c8df0d5241e840cd8bf9031a93 but I'm happy with the result. Rotation may generally effect on where exactly a sampling point ends up.

**Reply by vkrause** [reply]
> @vkrause I'd like to invite you to check out if the following would work for your different writer/generator use cases:
> 
> ```c++
> Barcode bc = CreateBarcodeFromText("Something something", BarcodeFormat::SomeFormat);
> Image img = bc.symbol(); // produces a 1 module == 1 pixel black/white image
> ```

Sorry for the delay, I have meanwhile adapted all our ZXing-using code to the latest API and things seem to work fine. The biggest annoyance was that the latest version still identifies as 2.3.0, making it non-straightforward to `ifdef` the new API, but that's a rather minor thing and obviously only temporary.

**Reply by axxel** [reply]
@vkrause sorry for the wrong version string (still have not learned that I should update the string immediately after a release, not before the next one...). Did you have a look at the new API for creating barcodes? Is that usable to you in this current form?

**Reply by vkrause** [reply]
The only part we use from the creation API is PDF417 with default settings I think (with text and binary content), I had no problems with the new API for that.

**Reply by axxel** [reply]
> The only part we use from the creation API is PDF417 with default settings I think (with text and binary content), I had no problems with the new API for that.

Very good to hear. I believe you mentioned quite a while back that you are using a number of other individual libraries to create other barcodes and might be interested in replacing them with zxing-cpp(zint). Is that still on the table?

**Reply by vkrause** [reply]
Right, that is still something to look into eventually. It's all similar though, just default settings for a few common formats (QR, Aztec, DataMatrix), with text and binary content, so I don't expect there to be any issues with that either. It's not a particularly high priority change for us though, as it will have very little user-facing impact nor save much maintenance effort.

**Comment by axxel** [comment]
Everyone: please have a look at my poll regarding the introduction of a string based 'variant' concept: #1025

**Reply by axxel** [reply]
**Update:** I ended up not introducing a variant (below the format) but a `symbology` (above the format). The new code is potentially breaking your code (if you use the C-API and use `ZXing_ReaderOptions_setFormat()` it is breaking your code). If you run into issues, let me know, I might add more deprecated helper code to keep the old API usable if reasonable.

Things to note:
* `BarcodeFormatFromString()` now throws if the string can not be parsed (same as `BarcodeFormatsFromString` always did)
* `ToString(BarcodeFromat)` now return human readable strings that are the official names (e.g. 'QR Code' instead of 'QRCode')
* Hence, `BarcodeFormatsFromString()` no longer supports ' ' as a separator character
* `ToString(BarcodeFormats)` now uses ", " instead of "|" as separators (the or-operator was a relic of the bit-field data storage) 

This is a preparation to potentially allow for the support of all formats supported be zint. This was some invasive changeset with a high probability of new bugs.

**Please** test this new code if you possibly can within a couple days, so the end of January deadline from Ubuntu can be kept.

**Comment by axxel** [comment]
Questions for the Python users (e.g. @Doekin @HamzaYslmn @chamalgomes @stumpylog @jameshilliard @nrlulz @fabulous-fabio @sayzzy-nt @edocod1 @anthonyryan1 @gf0507033 @sminkwitz @asesidaa @guylevin @eduardoatr @pathquester):

1. With the recent change of the `BarcodeFormats` implementation, I decided to change what the c++ function `ToString(BarcodeFormat)` returns, namely a standards conforming, human readable string (e.g. "QR Code" instead of "QRCode") and decided to reflect that in the `__str__` overload in python as well, meaning `print('{}', barcode.format)` now prints `QR Code` instead of `BarcodeFormat.QRCode`. Does anybody see an issue here?

2. There is a new `Barcode::extra()` property, which returns a JSON object serialized into a string. I see two options to implement that in python: a) simply pass on the JSON string as is or b) interpret that string inside the module and convert it into a standard python dictionary for direct access or even c) dynamically insert all key-value pairs from the JSON object directly into the `Barcode` object. Any strong preferences?

**Reply by HamzaYslmn** [reply]
1) If you are willing to accept a breaking change for string-based logic, the human-readable string is significantly better for the majority of use cases (display data etc).

2) Convert to Dictionary Option (b) because:
`barcode.extra['version']` is intuitive. `json.loads(barcode.extra)['version']` is clunky.
*   *Option (a) (String):* Too raw. It leaks implementation details (serialization format) to the user.
*   *Option (c) (Dynamic attributes):*
    *   It clutters the object namespace.
    *   It hampers IDE autocompletion and static analysis (Type hints won't know about these dynamic keys).
    *   It obscures where the data comes from (is it a core property or an extra?).

**Reply by axxel** [reply]
@HamzaYslmn thanks for your sound and reasoned opinion. I was leaning toward option (b) as well. Little followup question regarding most idiomatic python: if there is no extra data (c++ returns an empty string) should I rather return `None` or `{}`?

**Reply by HamzaYslmn** [reply]
> @HamzaYslmn thanks for your sound and reasoned opinion. I was leaning toward option (b) as well. Little followup question regarding most idiomatic python: if there is no extra data (c++ returns an empty string) should I rather return `None` or `{}`?

get.qrcode = data

if data:
 do this
else 
 do that

Should be "None" for easy use

**Reply by axxel** [reply]
Thanks but after some some research myself I would like to disagree: an empty dict evaluates to `False`, so your code works either way but an empty dict also allows to write `v = barcode.extra.get('Version')` which `None` does not. [Here](https://github.com/zxing-cpp/zxing-cpp/commit/2d7ceaf712d99a796225628ce999b76f533abb55) is the code.

**Reply by stumpylog** [reply]
Our own usage of the Python code is pretty basic.  If there's anything which does break, it will be east to fix.

My 2 cents though:

1. The string change is totally fine, we use it only for debug logging.
2. I would add a +1 for returning a dict, and and empty dict if there's no extra

**Comment by axxel** [comment]
I revisited the subject of performance again after quite a while. To my unpleasant (or pleasant) surprise, I managed to find several performance regressions (and a few new improvements) that lead to a total speed-up of the false positive use case (`./test/blackbox/ReaderTest samples -tfalsepositives`) of around 20% on my M4 ARM. If narrowed down to look only for linear barcodes, this can go up to a whooping 40% on specific input. On an Intel chip the gain will be at most half of that but still substantial.

Anyone interested in performing, please give the current HEAD a test run and let me know what you find. Also: did anyone notice the regression (most, if not all of it, introduced in 3.0)? If so: why did no-one say anything?

**Comment by axxel** [comment]
Just did a [3.1.0-rc1](https://github.com/zxing-cpp/zxing-cpp/releases/tag/v3.1.0-rc1) release. Please give it a test run if you can.

**Reply by axxel** [reply]
The new MicroPDF417 detector needed some considerable additional attention before I was happy with it. But here is the next release: [v3.1.0](https://github.com/zxing-cpp/zxing-cpp/releases/tag/v3.1.0).

**Reply by axxel** [reply]
Patch Release: [v3.1.1](https://github.com/zxing-cpp/zxing-cpp/releases/tag/v3.1.1).

**Comment by axxel** [comment]
I've added a new **[Commercial Support](Commercial%20Support.md)** page to the repository.

Many companies use `zxing-cpp` as part of commercial products. While the library will of course remain open source, maintaining and improving it takes a significant amount of time.

The new page describes the commercial support and consulting services that are available for companies that depend on `zxing-cpp` and would like direct access to the project's maintainer.

If your company uses `zxing-cpp`, I'd appreciate it if you could take a look. Whether through commercial support or GitHub Sponsors, every contribution helps ensure the project's continued maintenance and development.

Finally, a sincere thank you to everyone who already sponsors the project or has done so in the past. Your support has made a real difference.

**Comment by axxel** [comment]
To those who care about binary size of the library: see #1155 (How to deal with ECI/Unicode data when Unicode is disabled?)

**Comment by axxel** [comment]
I'm rewriting the black-box-texting code and accidentally stumbled over a very old "bug": calling `ReadBarcodes()` with `ReaderOptions::isPure(true)` always returns a `Barcodes` list with exactly one `Barcode`. That one object `!isValid()` when no Barcode is found. That behavior is plain wrong, it is supposed to return an empty `Barcodes` list (independent of the `isPure` flag).

I intend to simply fix this with 3.1.2. If someone wrote code that relies on the list never being empty, that code would then potentially crash after the update. Anyone sees any problem with this kind of change?

Note: `ReadBarcode()` (singular) always returns a single `Barcode` and that one might be invalid. So this is basically a copy'n'paste bug from when I introduced `ReadBarcodes()`.

**Reply by axxel** [reply]
Since I heard no objection: https://github.com/zxing-cpp/zxing-cpp/commit/ac627fb77f78e83c0df76feffe9005f671b1f2e2
